[Latoken] Add Latoken Exchange

roeiba opened the pull request.

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/service/LatokenTradeService.java:
```diff
@@ -0,0 +1,249 @@
+		
+    	try {
+    		if (params instanceof OpenOrdersParamCurrencyPair == false) {
+    			throw new CurrencyPairNotValidException("CurrencyPair is must be provided to get open orders.");
```
```
/**
 * Exception indicating that a request was made with a {@link CurrencyPair} that is not supported on
 * this exchange.
 *
 * @author bryant_harris
 */
public class CurrencyPairNotValidException extends ExchangeException {
```

not providing it in a method that requires it is a different think, just use `ExchangeException` here and below

roeiba commented:
Thank you for your comments, all polish is done.

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenAdapters.java:
```diff
@@ -0,0 +1,266 @@
+
+public class LatokenAdapters {
+
+  protected static List<LatokenPair> allPairs;
```
this type of static attributes will fail in multi threaded applications and should not be used
if you need to store some results from remoteInit then store it in an Exchange instance ad just pass it here every time you need it

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenAdapters.java:
```diff
@@ -0,0 +1,266 @@
+                        pair,
+                        null,
+                        null,
+                        BigDecimal.valueOf(level.getPrice())))
```
use builder

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenAdapters.java:
```diff
@@ -0,0 +1,266 @@
+                        pair,
+                        null,
+                        null,
+                        BigDecimal.valueOf(level.getPrice())))
```
use builder

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenAdapters.java:
```diff
@@ -0,0 +1,266 @@
+        pair,
+        BigDecimal.valueOf(latokenTrade.getPrice()),
+        latokenTrade.getTimestamp(),
+        null);
```
use builder

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenExchange.java:
```diff
@@ -0,0 +1,148 @@
+  private static final Logger LOG = LoggerFactory.getLogger(LatokenExchange.class);
+  private static final int PRECISION = 8;
+
+  public static final String sslUri = "https://api.latoken.com";
```
replace this with

```
private LatokenAuthenticated latoken;
```

roeiba replied:
I am not sure what you mean... Can you further explain?

walec51 replied:
please read my comments below and you'll understand why `String sslUri` is not needed here but `LatokenAuthenticated latoken` is

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenExchange.java:
```diff
@@ -0,0 +1,148 @@
+    this.marketDataService = new LatokenMarketDataService(this);
+    this.tradeService = new LatokenTradeService(this);
+    this.accountService = new LatokenAccountService(this);
+  }
```
replace this with

```
@Override
  protected void initServices() {
    latoken =
          RestProxyFactory.createProxy(LatokenAuthenticated.class, getExchangeSpecification().getSslUri(), getClientConfig());
    this.marketDataService = new LatokenMarketDataService(this, latoken);
    this.tradeService = new LatokenTradeService(this, latoken);
    this.accountService = new LatokenAccountService(this, latoken);
  }
```

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenExchange.java:
```diff
@@ -0,0 +1,148 @@
+      Map<CurrencyPair, CurrencyPairMetaData> pairsMetaData = exchangeMetaData.getCurrencyPairs();
+
+      Latoken latokenPublic =
+          RestProxyFactory.createProxy(LatokenAuthenticated.class, LatokenExchange.sslUri);
```
use the `latoken` I mentioned above

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenExchange.java:
```diff
@@ -0,0 +1,148 @@
+        CurrencyPairMetaData pairMetadata = LatokenAdapters.adaptPairMetaData(latokenPair);
+        addCurrencyPairMetadata(pairsMetaData, pair, pairMetadata);
+      }
+    } catch (LatokenException | IOException e) {
```
this catch seams redundant, the error will get logged twice, here and in the application code
just leave the second catch

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/LatokenExchange.java:
```diff
@@ -0,0 +1,148 @@
+            pairMetadata.getMinimumAmount(),
+            maxAmount,
+            pairMetadata.getPriceScale(),
+            null));
```
use builder

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/account/LatokenBalance.java:
```diff
@@ -0,0 +1,126 @@
+  private final String currencyId;
+  private final String symbol;
+  private final String name;
+  private final double amount;
```
use BigDecimal instead of double

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/exchangeinfo/LatokenCurrency.java:
```diff
@@ -0,0 +1,126 @@
+  private final String name;
+  private final int precision;
+  private final String type;
+  private final double fee;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/exchangeinfo/LatokenPair.java:
```diff
@@ -0,0 +1,174 @@
+  private final String symbol;
+  private final String baseCurrency;
+  private final String counterCurrency;
+  private final double makerFee;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/marketdata/LatokenOrderbook.java:
```diff
@@ -0,0 +1,121 @@
+
+  private final long pairId;
+  private final String symbol;
+  private final double spread;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/marketdata/LatokenTicker.java:
```diff
@@ -0,0 +1,158 @@
+
+  private final long pairId;
+  private final String symbol;
+  private final double volume;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/marketdata/LatokenTrade.java:
```diff
@@ -0,0 +1,88 @@
+public class LatokenTrade {
+
+  private final LatokenOrderSide side;
+  private final double price;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/marketdata/PriceLevel.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class PriceLevel {
+
+  private final double price;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/trade/LatokenNewOrder.java:
```diff
@@ -0,0 +1,156 @@
+  private final LatokenOrderSide side;
+  private final OrderSubclass type;
+  private final double price;
+  private final double amount;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/trade/LatokenOrder.java:
```diff
@@ -0,0 +1,132 @@
+
+  private final LatokenOrderStatus latokenOrderStatus;
+  private final double executedAmount;
+  private final double reaminingAmount;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/trade/LatokenOrderSide.java:
```diff
@@ -0,0 +1,17 @@
+    try {
+      return LatokenOrderSide.valueOf(s);
+    } catch (Exception e) {
+      throw new RuntimeException("Unknown LatokenOrderSide " + s + ".");
```
ExchangeException

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/trade/LatokenOrderStatus.java:
```diff
@@ -0,0 +1,20 @@
+    try {
+      return LatokenOrderStatus.valueOf(s);
+    } catch (Exception e) {
+      throw new RuntimeException("Unknown LatokenOrderStatus " + s + ".");
```
ExchangeException

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/trade/LatokenUserTrade.java:
```diff
@@ -0,0 +1,143 @@
+  private final double fee;
+  private final LatokenOrderSide side;
+  private final double price;
+  private final double amount;
```
use BigDecimal instead of dobule

walec51 commented on xchange-latoken/src/main/java/org/knowm/xchange/latoken/dto/trade/OrderSubclass.java:
```diff
@@ -0,0 +1,16 @@
+    try {
+      return OrderSubclass.valueOf(s);
+    } catch (Exception e) {
+      throw new RuntimeException("Unknown OrderSubclass " + s + ".");
```
ExchangeException

roeiba commented:
Thanks @walec51 !

I would love to hear your opinion on how to resolve a problem that I fixed in ugly manner:

On POST requests, latoken exchange expects not-empty body on the request. I added a dummy form parameter in order to force body creation by mazi. Can you suggest a cleaner solution?

walec51 commented:
> On POST requests, latoken exchange expects not-empty body on the request. I added a dummy form parameter in order to force body creation by mazi (see newOrder on LatokenAuthenticated).
> 
> Can you suggest a cleaner solution?

I don't think there is a cleaner solution then the one you proposed. Its ok with me

amjad1210 commented:
@roeiba Can you look into the failing test please?

All new PR are failing due to the following error.

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.929 s <<< FAILURE! - in org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest
[ERROR] org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest  Time elapsed: 0.929 s  <<< ERROR!
org.knowm.xchange.exceptions.ExchangeException: Failed to initialize: HTTP status code was not OK: 451
	at org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest.beforeClass(LatokenMarketDataServiceTest.java:29)
Caused by: si.mazi.rescu.HttpStatusIOException: HTTP status code was not OK: 451
	at org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest.beforeClass(LatokenMarketDataServiceTest.java:29)
```

roeiba commented:
Tests got broken due to latoken api update.

Will fix asap


On Tue, Nov 12, 2019 at 13:33 amjad1210 <notifications@github.com> wrote:

> @roeiba <https://github.com/roeiba> Can you look into the failing test
> please?
>
> All new PR are failing due to the following error.
>
> [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.929 s <<< FAILURE! - in org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest
> [ERROR] org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest  Time elapsed: 0.929 s  <<< ERROR!
> org.knowm.xchange.exceptions.ExchangeException: Failed to initialize: HTTP status code was not OK: 451
> 	at org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest.beforeClass(LatokenMarketDataServiceTest.java:29)
> Caused by: si.mazi.rescu.HttpStatusIOException: HTTP status code was not OK: 451
> 	at org.knowm.xchange.latoken.service.LatokenMarketDataServiceTest.beforeClass(LatokenMarketDataServiceTest.java:29)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/knowm/XChange/pull/3271?email_source=notifications&email_token=AASEN4MPHHHMUX52XBTEVXLQTKPBFA5CNFSM4JDNL472YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED2DDAA#issuecomment-552874368>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASEN4OBA5NHD7JUNOYAGEDQTKPBFANCNFSM4JDNL47Q>
> .
>

walec51 commented:
ok, also pls rename test classes that actually connect to exchanges so that they end with *Integration

for example rename LatokenMarketDataServiceTest to LatokenMarketDataIntegration

test classes that meet this naming criteria will be excluded from CI

walec51 commented:
I've renamed the test classes so @roeiba pls update your branch
now to run those tests you will have to set `skipIntegrationTests` to `false`

roeiba commented:
Just got to it, thank you for doing that and for the notification on the
tests.

Regards,
Roei


On Thu, Nov 14, 2019 at 12:52 PM Adam Walczak <notifications@github.com>
wrote:

> I've renamed the test classes so @roeiba <https://github.com/roeiba> pls
> update your branch
> now to run those tests you will have to set skipIntegrationTests to false
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/knowm/XChange/pull/3271?email_source=notifications&email_token=AASEN4LC7EYMSMAE2ZT4V23QTU3XXA5CNFSM4JDNL472YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEBSLNA#issuecomment-553854388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASEN4ON3D25HWUGRAEL4C3QTU3XXANCNFSM4JDNL47Q>
> .
>
